src/cowboy_http2.erl

```diff
@@ -238,8 +244,9 @@ loop(State=#state{parent=Parent, socket=Socket, transport=Transport,
 			setopts_active(State),
 			loop(State, Buffer);
 		%% System messages.
+		{'EXIT', Parent, shutdown} ->
+			loop(initiate_closing(State, shutdown), Buffer);
```

Comment:
With the above comments I suspect we'll need to add this clause to HTTP/1.1 as well. By the way we might also need to do this with `cowboy_websocket`. Perhaps this new shutdown clause should be added everywhere in a separate commit that explains why we are doing this, instead of together with the other changes.

Reply:
So you agree we want graceful shutdown for `shutdown` and not for other Parent exit reasons?

I think graceful shutdown for Parent shutdown is the main trigger of the Graceful shutdown feature, so it would be a pity to postpone that. I can add it for websocket as well if it's strait-forward.

Reply:
Btw, the exit reason after successful graceful shutdown will be `{shutdown, shutdown}` in this case, since terminate/2 wraps it like this. We could as well shutdown with a verbose tuple `{stop, {exit, shutdown}, 'Shutdown requested by parent process.'}` similar to the other cases. It's up to you.

In `cowboy_websocket`, the shutdown reasons are not verbose stop tuples like this.

Reply:
Just make `terminate` use `shutdown` when `Reason` is `shutdown`, and `{shutdown, Reason}` otherwise.

I don't think there's anything to postpone, I'm talking about a separate commit as part of this PR. Just to better show that these changes are changing more than just adding graceful shutdown. But both commits are part of the graceful shutdown feature. Alternatively be very explicit about this in the commit message.

Reply:
One more thing. This Reason passed to `initiate_closing/2` is only stored in timeout messages, not in the state. A graceful shutdown can end up terminating in `parse/2`, where we don't have the requested reason:

```Erlang
		more when Status =:= closing, Streams =:= #{} ->
			terminate(State, {stop, normal, 'The connection is going away.'});
```

We could either terminate with reason `normal` (which is also accepted by supervisor) or store the requested reason in the state to be able to use it here. If we want to consider the Reason passed to `sys:terminate/2,3`, we would have to do the latter.

We can also terminate in `down/3` (exit signal from children), which may also be a graceful shutdown:

```Erlang
	if
		State#state.http2_status =:= closing, State#state.streams =:= #{} ->
			terminate(State, {stop, normal, 'The connection is going away.'});
```

... or we can leave it as it is and forget about requested reason. What do you prefer?

Reply:
Leave those two cases as-is for now and let's see if that bothers anyone.

Reply:
OK. Btw, when I added a case for `shutdown` in `terminate/2`, Dialyzer tells me that `terminate(_, shutdown)` can never be called. It turns out it is always wrapped again in a stop tuple (e.g. `{stop, Reason, 'Graceful shutdown timed out.'}`) or replaced by another reason. Most often it seems (at least with Gun) that when shutdown is initiated, the client responds with GOAWAY and then closes the connection when all streams are done, so the Reason in terminate/2 will be `{stop, closed, 'The client is going away.'}`.

Because of this, I'm leaning towards skipping the shutdown with reason `shutdown`. If we're to write in a specific commit message that we shutdown with reason `shutdown` when requested by the supervisor, we should actually do that. Otherwise, there's no point. For now, I think it's better to stick to a stop tuple:

		{'EXIT', Parent, shutdown} ->
			Reason = {stop, {exit, shutdown}, 'Parent process requested shutdown.'}
			loop(initiate_closing(State, Reason), Buffer);

and then later, if it's important, rework this by storing the requested exit reason in the state.

Reply:
OK.